modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/ExecutionServiceImpl.java

```diff
@@ -674,16 +686,20 @@ private IgniteRel optimize(SqlNode sqlNode, IgnitePlanner planner) {
     }
 
     /** */
+    // TODO: prepareExplain should reuse prepare* methods for different query types
     private QueryPlan prepareExplain(SqlNode explain, PlanningContext ctx) throws ValidationException {
         IgnitePlanner planner = ctx.planner();
 
         SqlNode sql = ((SqlExplain)explain).getExplicandum();
 
         // Validate
-        explain = planner.validate(sql);
+        sql = planner.validate(sql);
 
         // Convert to Relational operators graph
-        IgniteRel igniteRel = optimize(explain, planner);
+        IgniteRel igniteRel = optimize(sql, planner);
+
+        if (sql.isA(ImmutableSet.of(SqlKind.INSERT, SqlKind.UPDATE)))
```

Comment:
Why logic here is differ from `prepareDml`? If we can't use physical plan, perhaps we can check top-level node of AST for `prepareDml` as it used here?

Reply:
Got it, `prepareDml` is executed only for insert, update and delete, so it's not so important there.

Reply:
Yes, `prepareExplain` is a separate branch that explains all types of other queries (regular query, DDL, DML).

I think I was wrong about putting shuttle to `prepare*` method. Looks like `optimize` is better place since it's a common point for both `prepareDml` and `prepareExplain`